static/js/services/enketo.js

```diff
@@ -258,19 +260,73 @@ angular.module('inboxServices').service('Enketo',
         });
     };
 
-    var storeXMLRecord = function(doc, record) {
+    function xpathish(e) {
+      for (var path = '', $e = $(e);
+          $e.length && !($e[0] instanceof Document);
+          $e = $e.parent()) {
+        path = '/' + $e[0].nodeName.toLowerCase() + path;
+      }
+      return path;
+    }
+
+    var xmlToDocs = function(doc, record) {
+      var docsToStore = [], idMap = {}, $record;
+
+      function mapOrAssignId(e) {
+        var id,
+            $id = $(e).children('_id');
+
+        if ($id.length) {
+          id = $id.text();
+        }
+
+        if (!id) {
+          id = uuid();
+        }
+
+        idMap[xpathish(e)] = id;
+      }
+
+      $record = $($($.parseXML(record)).children()[0]);
+      idMap[xpathish($record)] = doc._id || uuid();
+
+      $record.find('[db-doc=true]').each(function(i, e) {
+        mapOrAssignId(e);
+      });
+
+      $record.find('[doc-ref]').each(function(i, ref) {
+        var $ref = $(ref);
+        var refId = idMap[$ref.attr('doc-ref')];
+        $ref.text(refId);
+      });
+
+      $record.find('[db-doc=true]').each(function(i, e) {
+        var docToStore = EnketoTranslation.reportRecordToJs(e.outerHTML);
```

Comment:
No, can't merge blocks because the content of the XML has been updated between the two `each()` calls.  Could cache the result of the `find()` call if we wanted.